[JSC] Implement `Array.prototype.with` in C++

sosukesuzuki opened the pull request:
#### c0417a386c3062049eb1abec31004d7ecde8a4d1
<pre>
[JSC] Implement `Array.prototype.with` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=285241">https://bugs.webkit.org/show_bug.cgi?id=285241</a>

Reviewed by Yusuke Suzuki.

This patch implements `Array.prototype.with` in C++.

When FLT JIT is enabled, this patch makes it 1.1x ~ 1.2x faster.

                                         TipOfTree                  Patched

array-prototype-with-double           36.5286+-2.3897           32.1912+-4.1588          might be 1.1347x faster
array-prototype-with-int32            40.6655+-2.3469     ^     32.3867+-3.1772        ^ definitely 1.2556x faster
array-prototype-with-contiguous       35.9778+-3.8082           31.4018+-2.0459          might be 1.1457x faster

When DFG/FTL JIT is disabled, this patch makes it 1.6x ~ 1.8x faster.

                                         TipOfTree                  Patched

array-prototype-with-double           62.2153+-1.6448     ^     34.3870+-3.5752        ^ definitely 1.8093x faster
array-prototype-with-int32            64.1046+-3.0605     ^     36.2413+-8.5023        ^ definitely 1.7688x faster
array-prototype-with-contiguous       64.4120+-6.8815     ^     38.3409+-5.3097        ^ definitely 1.6800x faster
array-prototype-with-fast-path         0.4758+-0.3046            0.2803+-0.0861          might be 1.6977x faster

* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(toSpliced):
(with): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::argumentUnclampedIndexFromStartOrEnd):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:

Canonical link: <a href="https://commits.webkit.org/288360@main">https://commits.webkit.org/288360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0dd4604f0a5e3c662af8cfb977af71baeebb9a5a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36930 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87875 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33814 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10320 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64478 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22244 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85812 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1819 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75275 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44754 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1713 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29462 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32847 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75741 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72976 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30133 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89242 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81804 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10048 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7291 "Found 1 new test failure: http/wpt/mediarecorder/MediaRecorder-video-h264-profiles.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72892 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10276 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71083 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72107 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17916 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16270 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1436 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10001 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15522 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104209 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9875 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25257 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13340 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11644 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
Looks like some legit test failures are happening. Can you take a look?

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -192,6 +193,24 @@ static inline uint64_t argumentClampedIndexFromStartOrEnd(JSGlobalObject* global
+    double indexDouble = value.toIntegerOrInfinity(globalObject);
+    if (indexDouble < 0)
+        indexDouble += length;
+    return static_cast<int64_t>(indexDouble);
```
Casting infinity to `int64_t` is undefined behavior.

sosukesuzuki replied:
I've modified it so that if `indexDouble` is `Infinity`, it's approximated by the maximum or minimum `int64_t` value. What do you think?

Constellation replied:
That sounds fine to me!

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -605,6 +605,69 @@ JSArray* JSArray::fastToReversed(JSGlobalObject* globalObject, uint64_t length)
+
+        auto resultData = butterfly->contiguous().data();
+        memcpy(resultData, srcData, sizeof(JSValue) * length);
+
```
We need to set up to `vectorLength` with some valid values before setting it to `createWithButterfly`.

sosukesuzuki replied:
Thank you for your comment. I’m already calling `butterfly->setVectorLength(vectorLength);` before `createWithButterfly()`. Could you clarify what else needs to be done to address your concern?

Constellation replied:
I mean, you need to initialize fields of vector before `createWithButterfly`. Otherwise, concurrent GC will visit these vector's uninitialized fields and crash. Current code is only doing `memcpy` up to `length`. So `vectorLength - length` elements are uninitialized and leading to crash.

sosukesuzuki replied:
I got it! I've updated the code by referencing the `fastSlice` implementation.

sosukesuzuki committed:
[JSC] Implement `Array.prototype.with` in C++
https://bugs.webkit.org/show_bug.cgi?id=285241

Reviewed by Yusuke Suzuki.

This patch implements `Array.prototype.with` in C++.

When FLT JIT is enabled, this patch makes it 1.1x ~ 1.2x faster.

                                         TipOfTree                  Patched

array-prototype-with-double           36.5286+-2.3897           32.1912+-4.1588          might be 1.1347x faster
array-prototype-with-int32            40.6655+-2.3469     ^     32.3867+-3.1772        ^ definitely 1.2556x faster
array-prototype-with-contiguous       35.9778+-3.8082           31.4018+-2.0459          might be 1.1457x faster

When DFG/FTL JIT is disabled, this patch makes it 1.6x ~ 1.8x faster.

                                         TipOfTree                  Patched

array-prototype-with-double           62.2153+-1.6448     ^     34.3870+-3.5752        ^ definitely 1.8093x faster
array-prototype-with-int32            64.1046+-3.0605     ^     36.2413+-8.5023        ^ definitely 1.7688x faster
array-prototype-with-contiguous       64.4120+-6.8815     ^     38.3409+-5.3097        ^ definitely 1.6800x faster
array-prototype-with-fast-path         0.4758+-0.3046            0.2803+-0.0861          might be 1.6977x faster

* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(toSpliced):
(with): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::argumentUnclampedIndexFromStartOrEnd):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:

Canonical link: https://commits.webkit.org/288360@main
